src/inference.jl

```diff
@@ -24,25 +24,25 @@ end
 
 @kwdef struct TransformerLayerWeights
     # weights for rmsnorms
-    rms_att_weight::Vector{Float32} # (dim,)
-    rms_ffn_weight::Vector{Float32} # (dim,)
+    rms_att_weight::AbstractVector{Float32} # (dim,) # usually Float32
+    rms_ffn_weight::AbstractVector{Float32} # (dim,) # usually Float32
     # weights for matmuls
-    wq::Matrix # (dim, dim)
-    wk::Matrix # (dim, dim)
-    wv::Matrix # (dim, dim)
-    wo::Matrix # (dim, dim)
+    wq::AbstractMatrix # (dim, dim)
+    wk::AbstractMatrix # (dim, dim)
+    wv::AbstractMatrix # (dim, dim) # different quantization usually
+    wo::AbstractMatrix # (dim, dim)
     # weights for ffn
-    w1::Matrix # (dim, hidden_dim)
-    w2::Matrix # (hidden_dim, dim)
-    w3::Matrix # (dim, hidden_dim)
+    w1::AbstractMatrix # (dim, hidden_dim)
+    w2::AbstractMatrix # (hidden_dim, dim) # different quantization usually
+    w3::AbstractMatrix # (dim, hidden_dim)
```

Comment:
Could we run an MVP for this? Just to see if this really is needed? I would think you are right(I guess I have seen that), but not sure. 
Just because it would make some definitions a little bit more complex.

Reply:
I would consider it to be standard practice, and it shouldn't actually impact anything downstream. Most constructors will just work.

It's also hard to test trivial examples, we'd probably need to see benchmarks before and after. It's also entirely possible that this is a nothing change, but worth checking if you're interested.

Reply:
I see no real benefit to make these fields concretely-typed. Every layer can have differently quantized weights, and we do not want to compile new specialized methods for every combination of types. I deliberately made these fields abstractly-typed, to reduce the load on the compiler. I do not see how this would improve runtime performance.

Reply:
I wouldn't expect these types to have a massive number of different parametric types. For example, most of these will be either a GPU array or a CPU array, all of which are unlikely to change in value type and size. If it were the case that we were passing in a significant number of different types, then [yes it would be an issue](https://docs.julialang.org/en/v1/manual/performance-tips/#The-dangers-of-abusing-multiple-dispatch-(aka,-more-on-types-with-values-as-parameters)).

Imo the benefit of adding parametric types is to assist the compiler in 

1. Times of type instability. Parametric types can propagate upwards in the compile chain and help resolve periods where the output type would otherwise be moderately ambiguous. In this case, the compiler would observe that it should receive a `Matrix` or a `CuArray`. Without this, there is a chance that the compiler will only be able to specialize on `AbstractArray`. I don't know for sure that this is happening, and it is probably not, but including it is a "safe" option.
2. Precise function compilations. Most of these instances are heavy, large, and well lived, so compile time is probably a small part of the cost for these types. For heavy numeric operations like this, we need the most precise possible compiled operations, which is what you get from parametric types. 

In general I'm not sure how well tested the package is for type stability, but this would be an unambiguous way to force the compiler to tilt towards highly specific, well-optimized compiled code.

Up to you -- I personally think this is a very easy fix that should impact very little of the codebase. At best it gives you free speed and a higher baseline type stability. At worse you incur a very small and highly unlikely compile time cost in the event of volatile types used to construct these types.

Reply:
The weight arrays actually have do have different `eltype`s such as `block_q4_K`, `block_q5_K`, `block_q6_K`, etc. Having the fields in the weight structs abstractly typed just means that the matmul! calls in the inference function will do need a dynamic dispatch to identify the fully specialized kernel.
If we add type parameters to the `TransformerLayerWeights`, then the layers have to be stored in a heterogeneous array which also causes dynamic dispatches.

I agree though that `RunState` and `KVCache` should be parametrized -- since we are not running into the issues with the varying `eltype`s there.

Reply:
Sounds fair to me -- to clarify, that means this should be able to handle at most ~7 different matrix types and a two for the `rms` fields. This doesn't sound like too much to me.

One additional question -- how much do we actually expect those eltypes to change between constructions? My intuition would have been that it's quite rare to use a significant number of different block types that change between models. I would imagine that a model is mostly fixed to use the same types once it is instantiated, in which case the static dispatch is optimal.

My point there is that if you have a dispatchable amount of types, then it is probably optimal to provide the compiler the option to use static dispatch, especially since `matmul!` is called so many times. If those types are changing every second or millisecond or whatever, then yes, this is a stupid idea. My sense too is that we shouldn't expect a large number of layers. Llama 3 70b has 80 layers, and many of those are likely to have the same types. 

I think the idea here is that there is a trade off. Dynamic dispatch can accumulate a lot of overhead for high-frequency calls. If you have a bajillion types you would otherwise need to compile, then the dynamic dispatch makes sense. Neural nets are fixed once created, so we get something more closely representing a static DAG.

In the extreme case where you use homogeneous types, I'd hazard a guess that it's optimal to parametrically type, because you only ever compile one hot path.

The worst case scenario is a network of layers with infinite mix-and-match combinations of weights -- this would take a long time to compile but might perform better in general.

I'll drop it for now, since I think I've mentioned my views and ultimately it's a design decision by you. Appreciate you walking through it with me!

Reply:
I wouldn't think we need to argue about these. I pretty sure eventually going to have the time to benchmark both cases, bothfor compile time and runtime. There is no need to decide which would be more beneficial, we could also score how readable the 2 different code. :)

Reply:
I'm happy to move on. Just noting that this is a design consideration that may be worth thinking about for performance reasons.